internal/controller/grafanaorganization_controller.go

```diff
@@ -85,44 +83,112 @@ func (r *GrafanaOrganizationReconciler) Reconcile(ctx context.Context, req ctrl.
+		// We use the patch from sigs.k8s.io/cluster-api/util/patch to handle the patching without conflicts
+		logger.Info("adding finalizer", "finalizer", v1alpha1.GrafanaOrganizationFinalizer)
+		patchHelper, err := patch.NewHelper(grafanaOrganization, r.Client)
+		if err != nil {
 			return ctrl.Result{}, errors.WithStack(err)
 		}
+		controllerutil.AddFinalizer(grafanaOrganization, v1alpha1.GrafanaOrganizationFinalizer)
+		if err := patchHelper.Patch(ctx, grafanaOrganization); err != nil {
+			logger.Error(err, "failed to add finalizer", "finalizer", v1alpha1.GrafanaOrganizationFinalizer)
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+		logger.Info("added finalizer", "finalizer", v1alpha1.GrafanaOrganizationFinalizer)
+		return ctrl.Result{}, nil
 	}
 
 	// Ensure the first organization is renamed.
 	_, err := r.GrafanaAPI.Orgs.UpdateOrg(1, &grafanaAPIModels.UpdateOrgForm{
-		Name: sharedOrgName,
+		Name: grafana.SharedOrgName,
 	})
 	if err != nil {
-		logger.Error(err, fmt.Sprintf("Could not rename Main Org. to %s", sharedOrgName))
+		logger.Error(err, fmt.Sprintf("failed to rename Main Org. to %s", grafana.SharedOrgName))
 		return ctrl.Result{}, errors.WithStack(err)
 	}
 
-	//TODO Implement the logic to create the Grafana organization
+	// TODO add datasources for shared org.
+
+	// Create or update organization in Grafana
+	var organization grafana.Organization = grafana.Organization{
+		ID:   grafanaOrganization.Status.OrgID,
+		Name: grafanaOrganization.Spec.DisplayName,
+	}
 
+	if organization.ID == 0 {
+		// if the CR doesn't have an orgID, create the organization in Grafana
+		organization, err = grafana.CreateOrganization(ctx, r.GrafanaAPI, organization)
+	} else {
+		organization, err = grafana.UpdateOrganization(ctx, r.GrafanaAPI, organization)
+	}
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Update CR status if anything was changed
+	if organization.ID != grafanaOrganization.Status.OrgID {
+		grafanaOrganization.Status.OrgID = organization.ID
+
+		if err = r.Status().Update(ctx, grafanaOrganization); err != nil {
+			logger.Error(err, "failed to update the status")
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	}
+
+	// TODO add datasources for the organization.
+
+	err = r.configureGrafana(ctx)
+	if err != nil {
+		return ctrl.Result{}, errors.WithStack(err)
+	}
 	return ctrl.Result{}, nil
 }
 
 // reconcileDelete deletes the grafana organization.
 func (r GrafanaOrganizationReconciler) reconcileDelete(ctx context.Context, grafanaOrganization *v1alpha1.GrafanaOrganization) error {
 	logger := log.FromContext(ctx)
 
-	//TODO Implement the logic to delete the organization from Grafana.
+	// We do not need to delete anything if there is no finalizer on the grafana organization
+	if controllerutil.ContainsFinalizer(grafanaOrganization, v1alpha1.GrafanaOrganizationFinalizer) {
+
```

Comment:
Rather return early to avoid 1 unneeded indentation level
```suggestion
	if controllerutil.ContainsFinalizer(grafanaOrganization, v1alpha1.GrafanaOrganizationFinalizer) {
		return nil
	}

```

Reply:
You need to revert the condition right?